File: Task8-Akinator-with-interface/src/main/java/ru/thethenzou/akinator/Akinator.java
package ru.thethenzou.akinator;

import org.jpl7.*;
import org.jpl7.Integer;
import ru.thethenzou.models.Character;

import java.util.Optional;
import java.util.Properties;

/**
 * Класс для взаимодействия с акинатором написанным на прологе.
 */
public class Akinator {

    /**
     * Свойства.
     */
    Properties properties;

    /**
     * Конструктор.
     *
     * @param properties Свойства
     */
    public Akinator(Properties properties) {
        this.properties = properties;
        Query consult = new Query("consult('" + properties.getProperty("prolog.path") + "')");
        consult.hasMoreSolutions();
        loadPredicats();
    }

    private void loadPredicats() {
        Query loadQ1 = new Query("load_q1('" + properties.getProperty("q01") + "')");
        loadQ1.hasSolution();
        Query loadQ2 = new Query("load_q2('" + properties.getProperty("q02") + "')");
        loadQ2.hasSolution();
        Query loadQ4 = new Query("load_q4('" + properties.getProperty("q04") + "')");
        loadQ4.hasSolution();
        Query loadQ5 = new Query("load_q5('" + properties.getProperty("q05") + "')");
        loadQ5.hasSolution();
        Query loadQ6 = new Query("load_q6('" + properties.getProperty("q06") + "')");
        loadQ6.hasSolution();
        Query loadQ7 = new Query("load_q7('" + properties.getProperty("q07") + "')");
        loadQ7.hasSolution();
        Query loadQ8 = new Query("load_q8('" + properties.getProperty("q08") + "')");
        loadQ8.hasSolution();
        Query loadQ9 = new Query("load_q9('" + properties.getProperty("q09") + "')");
        loadQ9.hasSolution();
        Query loadQ10 = new Query("load_q10('" + properties.getProperty("q10") + "')");
        loadQ10.hasSolution();
        Query loadQ11 = new Query("load_q11('" + properties.getProperty("q11") + "')");
        loadQ11.hasSolution();
    }

    /**
     * Мето позволяет задать вопрос прологу.
     * @param character персонаж
     * @return optional с персонажем если он есть, а если нет то пустой
     */
    public Optional<Character> guess(Character character) {
        Variable characterName = new Variable("Name");
        Query guess = new Query("q", new Term[] {
                characterName,
                new Integer(character.getFirstQuestion()),
                new Integer(character.getSecondQuestion()),
                new Integer(character.getForthQuestion()),
                new Integer(character.getFifthQuestion()),
                new Integer(character.getSixthQuestion()),
                new Integer(character.getSeventhQuestion()),
                new Integer(character.getEighthQuestion()),
                new Integer(character.getNinethQuestion()),
                new Integer(character.getTenthQuestion()),
                new Integer(character.getElevenQuestion())
        });

        Optional<Character> answer;
        if (guess.hasSolution()) {
            character.setName(guess.oneSolution().get("Name").name());
            answer = Optional.of(character);
        } else {
            answer = Optional.empty();
        }
        return answer;
    }

    public void save(Character character) {
        Query guess = new Query("assert", new Term[] {
                new Atom(character.getName()),
                new Integer(character.getFirstQuestion()),
                new Integer(character.getSecondQuestion()),
                new Integer(character.getForthQuestion()),
                new Integer(character.getFifthQuestion()),
                new Integer(character.getSixthQuestion()),
                new Integer(character.getSeventhQuestion()),
                new Integer(character.getEighthQuestion()),
                new Integer(character.getNinethQuestion()),
                new Integer(character.getTenthQuestion()),
                new Integer(character.getElevenQuestion())
        });
        guess.hasSolution();

        Query saveQ1 = new Query("save_q1('" + properties.getProperty("q01") + "')");
        saveQ1.hasSolution();
        Query saveQ2 = new Query("save_q2('" + properties.getProperty("q02") + "')");
        saveQ2.hasSolution();
        Query saveQ4 = new Query("save_q4('" + properties.getProperty("q04") + "')");
        saveQ4.hasSolution();
        Query saveQ5 = new Query("save_q5('" + properties.getProperty("q05") + "')");
        saveQ5.hasSolution();
        Query saveQ6 = new Query("save_q6('" + properties.getProperty("q06") + "')");
        saveQ6.hasSolution();
        Query saveQ7 = new Query("save_q7('" + properties.getProperty("q07") + "')");
        saveQ7.hasSolution();
        Query saveQ8 = new Query("save_q8('" + properties.getProperty("q08") + "')");
        saveQ8.hasSolution();
        Query saveQ9 = new Query("save_q9('" + properties.getProperty("q09") + "')");
        saveQ9.hasSolution();
        Query saveQ10 = new Query("save_q10('" + properties.getProperty("q10") + "')");
        saveQ10.hasSolution();
        Query saveQ11 = new Query("save_q11('" + properties.getProperty("q11") + "')");
        saveQ11.hasSolution();

        loadPredicats();
    }
}

File: Task8-Akinator-with-interface/src/main/resources/akinator.pl
read_str(A):-get0(X),r_str(X,A,[]).
r_str(10,A,A):-!.
r_str(X,A,B):-append(B,[X],B1),get0(X1),r_str(X1,A,B1).

q01('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q1.txt').
q02('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q2.txt').
q04('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q4.txt').
q05('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q5.txt').
q06('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q6.txt').
q07('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q7.txt').
q08('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q8.txt').
q09('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q9.txt').
q10('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q10.txt').
q11('E:/My_Progs/knowlage-base/Task8-Akinator-with-interface/src/main/resources/q11.txt').

% =============================================================================
%Question 1
:- dynamic q1/2.

write_q1(X,Y):-
    repeat,
    (q1(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q1(X,Y)));
    X=nil,
    Y=nil).

read_q1(-1):-!.
read_q1(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q1(X,Y)),
    get0(Sym),
    read_q1(Sym).


save_q1(File):-
    tell(File),
    write_q1(X,_),
    X=nil,
    told.

load_q1(File):-
    see(File),
    get0(Sym),
    read_q1(Sym),
    seen.

question1(X1):-
    write("Какого пола ваш персонаж? "),nl,
    write("1. Мужской"),nl,
    write("2. Женский"),nl,
    write("3. Другой"),nl,
    read(X1).


% =============================================================================
%Question 2
:- dynamic q2/2.

write_q2(X,Y):-
    repeat,
    (q2(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q2(X,Y)));
    X=nil,
    Y=nil).

read_q2(-1):-!.
read_q2(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q2(X,Y)),
    get0(Sym),
    read_q2(Sym).


save_q2(File):-
    tell(File),
    write_q2(X,_),
    X=nil,
    told.

load_q2(File):-
    see(File),
    get0(Sym),
    read_q2(Sym),
    seen.


question2(X2):-
    write("Ваш персонаж человек?"),nl,
    write("1. Да"),nl,
    write("2. Нет"),nl,
    write("3. Не знаю"),nl,
    read(X2).


% =============================================================================
%Question 4
:- dynamic q4/2.

write_q4(X,Y):-
    repeat,
    (q4(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q4(X,Y)));
    X=nil,
    Y=nil).

read_q4(-1):-!.
read_q4(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q4(X,Y)),
    get0(Sym),
    read_q4(Sym).


save_q4(File):-
    tell(File),
    write_q4(X,_),
    X=nil,
    told.

load_q4(File):-
    see(File),
    get0(Sym),
    read_q4(Sym),
    seen.


question4(X4):-
    write("Ваш персонаж умеет летать?"),nl,
    write("1. Да"),nl,
    write("2. Нет"),nl,
    write("3. Не знаю"),nl,
    read(X4).


% =============================================================================
%Question 5
:- dynamic q5/2.

write_q5(X,Y):-
    repeat,
    (q5(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q5(X,Y)));
    X=nil,
    Y=nil).

read_q5(-1):-!.
read_q5(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q5(X,Y)),
    get0(Sym),
    read_q5(Sym).


save_q5(File):-
    tell(File),
    write_q5(X,_),
    X=nil,
    told.

load_q5(File):-
    see(File),
    get0(Sym),
    read_q5(Sym),
    seen.


question5(X5):-
    write("Ваш персонаж владеет магией?"),nl,
    write("1. Да"),nl,
    write("2. Нет"),nl,
    write("3. Не знаю"),nl,
    read(X5).


% =============================================================================
%Question 6
:- dynamic q6/2.

write_q6(X,Y):-
    repeat,
    (q6(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q6(X,Y)));
    X=nil,
    Y=nil).

read_q6(-1):-!.
read_q6(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q6(X,Y)),
    get0(Sym),
    read_q6(Sym).


save_q6(File):-
    tell(File),
    write_q6(X,_),
    X=nil,
    told.

load_q6(File):-
    see(File),
    get0(Sym),
    read_q6(Sym),
    seen.


question6(X6):-
    write("Владеет ли ваш персонаж боевыми искусствами?"),nl,
    write("1. Да"),nl,
    write("2. Нет"),nl,
    write("3. Не знаю"),nl,
    read(X6).


% =============================================================================
%Question 7
:- dynamic q7/2.

write_q7(X,Y):-
    repeat,
    (q7(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q7(X,Y)));
    X=nil,
    Y=nil).

read_q7(-1):-!.
read_q7(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q7(X,Y)),
    get0(Sym),
    read_q7(Sym).


save_q7(File):-
    tell(File),
    write_q7(X,_),
    X=nil,
    told.

load_q7(File):-
    see(File),
    get0(Sym),
    read_q7(Sym),
    seen.


question7(X7):-
    write("У вашего персонажа четыре руки? "),nl,
    write("1. Да"),nl,
    write("2. Нет"),nl,
    read(X7).


% =============================================================================
%Question 8
:- dynamic q8/2.

write_q8(X,Y):-
    repeat,
    (q8(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q8(X,Y)));
    X=nil,
    Y=nil).

read_q8(-1):-!.
read_q8(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q8(X,Y)),
    get0(Sym),
    read_q8(Sym).


save_q8(File):-
    tell(File),
    write_q8(X,_),
    X=nil,
    told.

load_q8(File):-
    see(File),
    get0(Sym),
    read_q8(Sym),
    seen.


question8(X8):-
    write("Сколько лет вашему персонажу? "),nl,
    write("1. Меньше 30 лет"),nl,
    write("2. Mеньше 55 лет"),nl,
    write("3. Больше 55 ле"),nl,
    write("4. Да"),nl,
    write("5. Не знаю"),nl,
    read(X8).


% =============================================================================
%Question 9
:- dynamic q9/2.

write_q9(X,Y):-
    repeat,
    (q9(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q9(X,Y)));
    X=nil,
    Y=nil).

read_q9(-1):-!.
read_q9(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q9(X,Y)),
    get0(Sym),
    read_q9(Sym).


save_q9(File):-
    tell(File),
    write_q9(X,_),
    X=nil,
    told.

load_q9(File):-
    see(File),
    get0(Sym),
    read_q9(Sym),
    seen.


question9(X9):-
    write("Ваш персонаж из DLC?"),nl,
    write("1. Да"),nl,
    write("2. Нет"),nl,
    read(X9).


% =============================================================================
%Question 10
:- dynamic q10/2.

write_q10(X,Y):-
    repeat,
    (q10(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q10(X,Y)));
    X=nil,
    Y=nil).

read_q10(-1):-!.
read_q10(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q10(X,Y)),
    get0(Sym),
    read_q10(Sym).


save_q10(File):-
    tell(File),
    write_q10(X,_),
    X=nil,
    told.

load_q10(File):-
    see(File),
    get0(Sym),
    read_q10(Sym),
    seen.


question10(X10):-
    write("Пользуется ли Ваш персонаж оружием?"),nl,
    write("1. Да"),nl,
    write("2. Нет"),nl,
    read(X10).


% =============================================================================
%Question 11
:- dynamic q11/2.

write_q11(X,Y):-
    repeat,
    (q11(X,Y) -> (
         nl,
         write(X),
         nl,
         write(Y),
         write("."),
         retract(q11(X,Y)));
    X=nil,
    Y=nil).

read_q11(-1):-!.
read_q11(_):-
    read_str(Lang),
    name(X,Lang),
    read(Y),
    asserta(q11(X,Y)),
    get0(Sym),
    read_q11(Sym).


save_q11(File):-
    tell(File),
    write_q11(X,_),
    X=nil,
    told.

load_q11(File):-
    see(File),
    get0(Sym),
    read_q11(Sym),
    seen.


question11(X11):-
    write("Какое мировозрение у Вашего персонажа?"),nl,
    write("1. Добро"),nl,
    write("2. Зло"),nl,
    write("3. Нейтральное"),nl,
    write("4. Не знаю"),nl,
    read(X11).


% =============================================================================
%run

load:-
    q01(File1),q02(File2),q04(File4),q05(File5),q06(File6),
    q07(File7),q08(File8),q09(File9),q10(File10),q11(File11),
    load_q1(File1),load_q2(File2),load_q4(File4),load_q5(File5),load_q6(File6),
    load_q7(File7),load_q8(File8),load_q9(File9),load_q10(File10),load_q11(File11).

save:-
    q01(File1),q02(File2),q04(File4),q05(File5),q06(File6),
    q07(File7),q08(File8),q09(File9),q10(File10),q11(File11),
    save_q1(File1),save_q2(File2),save_q4(File4),save_q5(File5),save_q6(File6),
    save_q7(File7),save_q8(File8),save_q9(File9),save_q10(File10),save_q11(File11).

question(X1,X2,X4,X5,X6,X7,X8,X9,X10,X11):-
    question1(X1),question2(X2),question4(X4),question5(X5),
    question6(X6),question7(X7),question8(X8),question9(X9),
    question10(X10),question11(X11),get0(_).

q(X,X1,X2,X4,X5,X6,X7,X8,X9,X10,X11):-
    q1(X,X1),q2(X,X2),q4(X,X4),q5(X,X5),q6(X,X6),
    q7(X,X7),q8(X,X8),q9(X,X9),q10(X,X10),q11(X,X11).

assert(Name,X1,X2,X4,X5,X6,X7,X8,X9,X10,X11):-
    asserta(q1(Name,X1)),asserta(q2(Name,X2)),
    asserta(q4(Name,X4)),asserta(q5(Name,X5)),
    asserta(q6(Name,X6)),asserta(q7(Name,X7)),
    asserta(q8(Name,X8)),asserta(q9(Name,X9)),
    asserta(q10(Name,X10)),asserta(q11(Name,X11)).

run:-
    load,
    question(X1,X2,X4,X5,X6,X7,X8,X9,X10,X11),
    (q(X,X1,X2,X4,X5,X6,X7,X8,X9,X10,X11) ->
    write(X);

    writeln("Write your answer:"),
    read_str(Answer),
    name(Name,Answer),
    assert(Name,X1,X2,X4,X5,X6,X7,X8,X9,X10,X11),
    save).

